game_patch/multi/server.cpp

```diff
@@ -305,6 +316,36 @@ static void send_private_message_with_stats(rf::Player* player)
     send_chat_line_packet(str.c_str(), player);
 }
 
+void shuffle_level_array()
+{
+    std::size_t n = rf::netgame.levels.size();
+    for (std::size_t i = n - 1; i > 0; --i) {
+        std::size_t j = dist(rng) % (i + 1);
+        std::swap(rf::netgame.levels[i], rf::netgame.levels[j]);
+    }
+    xlog::info("Shuffled level rotation");
+}
+
+const char* get_rand_level_filename()
+{
+    int num_levels = rf::netgame.levels.size();
+
+    if (num_levels > 1) {
+        std::uniform_int_distribution<int> dist(0, num_levels - 2);
+        int rand_level_index = dist(rng);
+
+        // avoid selecting the current level
+        if (rand_level_index >= rf::netgame.current_level_index) {
+            rand_level_index++;
+        }
+
+        return rf::netgame.levels[rand_level_index].c_str();
+    }
+
+    // nowhere else to go, we're staying here!
+    return rf::netgame.levels[rf::netgame.current_level_index].c_str();
```

Comment:
Consider `std::sample`. It can take `rng` as a parameter.

Reply:
From everything I'm seeing, while `std::sample` would make more sense if it was grabbing multiple entries, the current code is more efficient, readable, and simplified than using `std::sample` would be given we only need to grab 1 entry.

Reply:
The following code has simpler logic imo.

```C++
    int num_levels = rf::netgame.levels.size();

    if (num_levels > 1) {
        auto level_indices = std::views::iota(0, num_levels) 
                | std::views::filter([] (int i) { return i != rf::netgame.current_level_index; });

        int rand_level_index{};
        std::ranges::sample(level_indices, &rand_level_index, 1, rng);

        return rf::netgame.levels[rand_level_index].c_str();
    }

    // nowhere else to go, we're staying here!
    return rf::netgame.levels[rf::netgame.current_level_index].c_str();
```

Reply:
On second thought I prefer yours. However usually it is better to return early then indent which I think is the case here. Further why is `rand_level_index >= rf::netgame.current_level_index` not using `==`?

```C++
    int num_levels = rf::netgame.levels.size();

    if (num_levels <= 1) {
        // nowhere else to go, we're staying here!
        return rf::netgame.levels[rf::netgame.current_level_index].c_str();
    }

    std::uniform_int_distribution dist{0, num_levels - 2};
    int rand_level_index = dist(rng);

    // avoid selecting the current level
    if (rand_level_index >= rf::netgame.current_level_index) {
        ++rand_level_index;
    }

    return rf::netgame.levels[rand_level_index].c_str();
```

Reply:
> Further why is `rand_level_index >= rf::netgame.current_level_index` not using `==`?

If it were using `==`, the last level could never be returned.

`std::uniform_int_distribution<int> dist(0, num_levels - 2);` excludes the last index before selecting one randomly, so if the randomly selected index is either the current level _or_ a later one, we need to shift ahead by 1 to make it possible for the last index to be selected.